pkg/client/interceptor/interceptor.go

```diff
@@ -2,5 +2,320 @@
+
+			return streamErr
+		})
+		if err != nil {
+			return nil, err
+		}
+
+		return stream, nil
+	}
+}
+
+func (i *Interceptor) intercept(ctx context.Context, cc *grpc.ClientConn, method string, fn func(context.Context) error) error {
+	if ctx.Value(SkipInterceptorContextKey{}) != nil {
+		return fn(ctx)
+	}
+
+	ctx = context.WithValue(ctx, SkipInterceptorContextKey{}, struct{}{})
+
+	if err := i.initializeOnce(ctx, cc); err != nil {
+		return err
+	}
+
+	if !i.authEnabled {
+		return fn(ctx)
+	}
+
+	unsignedCtx := ctx
+	isRetryable := i.serviceAccount == nil
+
+	signAndMakeCall := func() (bool, error) {
+		signedCtx, err := i.sign(unsignedCtx, cc, method)
+		if err != nil {
+			return isRetryable, err
+		}
+
+		err = fn(signedCtx)
+		if err != nil {
+			return status.Code(err) == codes.Unauthenticated && isRetryable, err
+		}
+
+		return false, nil
+	}
+
+	for {
+		retry, err := signAndMakeCall()
+		if err == nil { // call succeeded
+			return nil
+		}
+
+		if !retry { // should not retry
+			return err
+		}
+
+		fmt.Fprintf(i.options.InfoWriter, "Could not authenticate: %v\n", err)
+
+		if err = i.renewUser(ctx, cc); err != nil {
+			return err
+		}
+
+		isRetryable = false // mark as not retryable since we already tried once
+	}
+}
+
+func (i *Interceptor) renewUser(ctx context.Context, cc *grpc.ClientConn) error {
+	newSigner, err := i.options.RenewUserKeyFunc(ctx, cc, &i.options)
+	if err != nil {
+		return err
+	}
+
+	i.setUserSigner(newSigner)
+
+	return nil
+}
+
+func (i *Interceptor) sign(ctx context.Context, cc *grpc.ClientConn, method string) (context.Context, error) {
+	md, ok := metadata.FromOutgoingContext(ctx)
+	if !ok {
+		md = metadata.New(nil)
+	}
+
+	msg := message.NewGRPC(md, method)
+
+	var (
+		identity string
+		signer   message.Signer
+	)
+
+	if i.serviceAccount != nil {
+		identity = i.serviceAccount.Name
+		signer = i.serviceAccount.Key
+	} else {
+		identity = i.options.Identity
+
+		var err error
+
+		signer, err = i.initAndGetUserSigner(ctx, cc)
+		if err != nil {
+			return nil, fmt.Errorf("failed to get signer: %w", err)
+		}
+	}
+
+	if err := msg.Sign(identity, signer); err != nil {
+		return nil, fmt.Errorf("failed to sign message: %w", err)
+	}
+
+	return metadata.NewOutgoingContext(ctx, msg.Metadata), nil
+}
+
+func (i *Interceptor) initAndGetUserSigner(ctx context.Context, cc *grpc.ClientConn) (message.Signer, error) {
+	i.userSignerLock.Lock()
+	defer i.userSignerLock.Unlock()
+
+	if i.userSigner != nil {
+		return i.userSigner, nil
+	}
+
+	key, err := i.options.GetUserKeyFunc(ctx, cc, &i.options)
+	if err != nil {
+		return nil, err
+	}
+
+	i.userSigner = key
+
+	return i.userSigner, nil
+}
+
+func (i *Interceptor) setUserSigner(signer message.Signer) {
+	i.userSignerLock.Lock()
+	defer i.userSignerLock.Unlock()
+
+	i.userSigner = signer
+}
+
+func (i *Interceptor) initializeOnce(ctx context.Context, cc *grpc.ClientConn) error {
+	var err error
+
+	i.initOnce.Do(func() {
```

Comment:
Looks better, but now with this code if we try to call `initializeOnce` second time it will return no error.
Maybe we can just have a lock and try to init, until it succeeds, instead of using Once?

Reply:
oh right 🤦 that was the reason why I did that with `initErr`.

I think `once` is fine, because I want to have the behavior of once - so even if I use a lock, I'd reinvent once behavior.

What do you think about it now - I just brought back initErr but there's only 1 place it gets assigned a value, and `initializeOnce` returns an error still. So the flow is easier to read and its behavior is as expected.